python/pyarrow/tests/test_fs.py

```diff
@@ -195,6 +193,58 @@ def test_cannot_instantiate_base_filesystem():
         FileSystem()
 
 
+def test_filesystem_equals():
+    fs0 = LocalFileSystem()
+    fs1 = LocalFileSystem()
+    fs2 = _MockFileSystem()
+
+    assert fs0.equals(fs0)
+    assert fs0.equals(fs1)
+    with pytest.raises(TypeError):
+        fs0.equals('string')
+    assert fs0 == fs0 == fs1
+    assert fs0 != 4
+
+    assert fs2 == fs2
+    assert fs2 != _MockFileSystem()
+
+    assert SubTreeFileSystem('/base', fs0) == SubTreeFileSystem('/base', fs0)
+    assert SubTreeFileSystem('/base', fs0) != SubTreeFileSystem('/base', fs2)
+    assert SubTreeFileSystem('/base', fs0) != SubTreeFileSystem('/other', fs0)
+
+
+def test_filesystem_pickling(fs):
+    if isinstance(fs, _MockFileSystem):
+        pytest.xfail(reason='MockFileSystem is not serializable')
+
+    serialized = pickle.dumps(fs)
+    restored = pickle.loads(serialized)
+    assert isinstance(restored, FileSystem)
+    assert restored.equals(fs)
+
+
+def test_filesystem_is_functional_after_pickling(fs, pathfn):
+    if isinstance(fs, _MockFileSystem):
+        pytest.xfail(reason='MockFileSystem is not serializable')
+
+    restored = pickle.loads(pickle.dumps(fs))
+
+    aaa = pathfn('a/aa/aaa/')
+    bb = pathfn('a/bb')
+    c = pathfn('c.txt')
+
+    restored.create_dir(aaa)
```

Comment:
Rather, should create the files using `fs` and inspect them using `restored`, to make sure the same storage is being accessed.